### PR TITLE
Mlue 9676

### PR DESCRIPTION
**Pull request by marianogonzalez**

### CODE REVIEW

File: extensions/ftp/src/main/java/org/mule/extension/ftp/api/ftp/FtpFileSystem.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import org.mule.runtime.api.connection.ConnectionValidationResult;
+import org.mule.runtime.module.extension.file.api.FileAttributes;
+import org.mule.runtime.module.extension.file.api.FileSystem;
+
+import java.io.InputStream;
+
+/**
+ * Specialization of the {@link FileSystem} contract for
+ * file systems mounted on FTP/SFTP servers
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/api/ftp/FtpFileSystem.java]: Isn't this class only for FTP?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Path;
+
+import javax.inject.Inject;
+
+import org.apache.commons.net.ftp.FTPClient;
+
+/**
+ * Implementation of {@link FtpFileSystem} for files residing on a SFTP server
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java]: I think you need to include the package for the link to work in the javadoc.

File: extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    protected PathLock createLock(Path path, Object... params)
+    {
+        return new URLPathLock(toURL(path), muleContext.getLockFactory());
+    }
+
+    private URL toURL(Path path)
+    {
+        try
+        {
+            return new URL("ftp", client.getHost(), client.getPort(), path != null ? path.toString() : EMPTY);
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java]: Do you have a constant elsewhere for "ftp"?. If so use it here.

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+            {
+                fileSystem.delete(targetFile.getPath());
+            }
+            else
+            {
+                throw command.exception(String.format("Cannot copy file '%s' to path '%s' because it already exists", source.getPath(), target));
+            }
+        }
+
+        try (InputStream inputStream = fileSystem.retrieveFileContent(source))
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java]: Could it be the case that the InputStream is already in the payload in this scenario?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        {
+            if (inputStream == null)
+            {
+                throw command.exception(String.format("Could not read file '%s' while trying to copy it to remote path '%s'", source.getPath(), target));
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java]: static imports

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/ClassicFtpFileAttributes.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import java.time.LocalDateTime;
+
+import org.apache.commons.net.ftp.FTPFile;
+
+/**
+ * Implementation of {@link FtpFileAttributes} for files read from a FTP server.
+ *
+ * @since 4.0
+ */
+public class ClassicFtpFileAttributes extends AbstractFileAttributes implements FtpFileAttributes
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/ClassicFtpFileAttributes.java]: What's the meaning of classic? Can you think of a more meaningful word for it?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/ClassicFtpCommand.java [hunk 7]
```diff
@@ -121,33 +91,33 @@ private FtpFileAttributes getFile(String filePath, boolean requireExistence)
...
+        Stack<Path> fragments = new Stack<>();
         try
         {
-            changeWorkingDirectory("/");
-            for (int i = 0; i < directoryPath.getNameCount(); i++)
+            for (int i = directoryPath.getNameCount(); i >= 0; i--)
             {
-                String fragment = directoryPath.getName(i).toString();
-                if (!tryChangeWorkingDirectory(fragment))
+                Path subPath = Paths.get("/").resolve(directoryPath.subpath(0, i));
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/ClassicFtpCommand.java]: Should this be File.pathSeparator?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+     * Template method that renames the file at {@code filePath} to {@code newName}.
+     * <p>
+     * This method performs path resolution and validation and eventually delegates
+     * into {@link #doRename(String, String)}, in which the actual renaming implementation
+     * is.
+     *
+     * @param filePath the path of the file to be renamed
+     * @param newName  the new name
+     */
+    protected void rename(String filePath, String newName)
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java]: Should this operation take into account the override capabilities?

Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java]: Should this method be final to make sure all the extensions have the same behaviour?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+     * @param filePath the path of the file to be renamed
+     * @param newName  the new name
+     * @throws Exception if anything goes wrong
+     */
+    protected abstract void doRename(String filePath, String newName) throws Exception;
+
+    /**
+     * {@inheritDoc}
+     */
+    protected void createDirectory(String basePath, String directoryName)
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java]: Override functionality should apply?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpConnectionProvider.java [hunk 10]
```diff
@@ -44,58 +52,35 @@
...
+     * <p>
+     * Defaults to {@code BINARY}
      */
-    @Override
-    public FtpFileSystem connect(FtpConnector config) throws ConnectionException
-    {
-        return new FtpFileSystem(config, createClient(config));
-    }
+    @Parameter
+    @Optional(defaultValue = "BINARY")
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpConnectionProvider.java]: If the value is an enum, shouldn't the defaultValue also be an enum?

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpFileSystem.java [hunk 11]
```diff
@@ -246,23 +268,22 @@ protected PathLock createLock(Path path, Object... params)
 
     private URL toURL(Path path)
     {
-        URL url;
         try
         {
-            url = new URL("ftp", client.getRemoteAddress().toString(), client.getRemotePort(), path != null ? path.toString() : EMPTY);
+            return new URL("ftp", client.getRemoteAddress().toString(), client.getRemotePort(), path != null ? path.toString() : EMPTY);
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpFileSystem.java]: Use same constant extracted from the previous comment.

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/sftp/connection/SftpConnectionProvider.java [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+import org.mule.runtime.extension.api.annotation.param.Optional;
+import org.mule.runtime.extension.api.annotation.param.display.Password;
+
+/**
+ * An {@link AbstractFtpConnectionProvider} which provides instances of
+ * {@link SftpFileSystem} from instances of {@link FtpConnector}
+ *
+ * @since 4.0
+ */
+@Alias("sftp")
```
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/sftp/connection/SftpConnectionProvider.java]: Extract constant if used elsewhere

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpCopyTestCase.java [hunk 13]
```diff
@@ -133,12 +141,12 @@ public void directoryAndOverwrite() throws Exception
...
-        ftpClient.makeDir(target);
-        ftpClient.changeWorkingDirectory(target);
-        ftpClient.makeDir(SOURCE_DIRECTORY_NAME);
-        ftpClient.putFile(SOURCE_FILE_NAME, SOURCE_DIRECTORY_NAME, EXISTING_CONTENT);
+        testHarness.makeDir(target);
+        testHarness.changeWorkingDirectory(target);
+        testHarness.makeDir(SOURCE_DIRECTORY_NAME);
+        testHarness.write(SOURCE_DIRECTORY_NAME, SOURCE_FILE_NAME, EXISTING_CONTENT);
 
-        ftpClient.changeWorkingDirectory("../");
```
Comment (pablolagreca) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpCopyTestCase.java]: File.separator?

File: modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileContentWrapper.java [hunk 14]
```diff
@@ -130,7 +130,7 @@ private Object tryTransform(DataType sourceDataType, DataType targetDataType)
         Transformer transformer;
         try
         {
-            transformer = event.getMuleContext().getRegistry().lookupTransformer(sourceDataType, targetDataType);
+            transformer = castEvent().getMuleContext().getRegistry().lookupTransformer(sourceDataType, targetDataType);
```
Comment (pablolagreca) [inline modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileContentWrapper.java]: Can you add the MuleContext as a constructor parameter of this class instead?. We will likely remove the MuleContext from the MuleEvent.

File: modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileSystem.java [hunk 15]
```diff
@@ -219,6 +220,8 @@ void write(String filePath,
      */
     PathLock lock(Path path, Object... params);
 
+    Lock createMuleLock(String id);
```
Comment (pablolagreca) [inline modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileSystem.java]: Why would this class expose a createMuleLock method?. Makes sense to use it internally but it should not be exposed to end users.

File: extensions/ftp/src/main/java/org/mule/extension/ftp/api/ftp/FtpFileSystem.java (hunk 1 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/api/ftp/FtpFileSystem.java]: this is a base interface, common to both

File: extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java (hunk 2 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java]: no, it also works if you have an import to that class. It's actually what intelliJ recommends you to do

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java (hunk 4 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java]: copy and move work in terms of paths. If you have the input stream in the payload then it makes no sense to use copy, you would just write to a different location

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java (hunk 5 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/AbstractFtpCopyDelegate.java]: done

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/ClassicFtpFileAttributes.java (hunk 6 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/ClassicFtpFileAttributes.java]: I meant normal, regular, good old fashioned FTP. That's the best name I could come out with, if you have a better idea I'm open to it

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/ClassicFtpCommand.java (hunk 7 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/ClassicFtpCommand.java]: no. FTP uses `/` even if windows sets it to `\`

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java (hunk 8 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java]: i'll make it final. as for overwrite that would also affect the behavior of file so I would do it in a separate PR

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java (hunk 9 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java]: no. it should fail IMO

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpConnectionProvider.java (hunk 10 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpConnectionProvider.java]: Java annotations don't allow that. This is the only way of doing it, SDK will do the proper transformations transparently

File: extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java (hunk 3 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/api/sftp/SftpFileSystem.java]: This is actually a bug. It should be sftp

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpFileSystem.java (hunk 11 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/connection/ClassicFtpFileSystem.java]: look at the response on the other comment

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/sftp/connection/SftpConnectionProvider.java (hunk 12 above)
Comment (marianogonzalez) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/sftp/connection/SftpConnectionProvider.java]: this alias should not match the constant you mention. If now happens to match, but they shouldn't be tied together

File: extensions/ftp/src/test/java/org/mule/extension/ftp/FtpCopyTestCase.java (hunk 13 above)
Comment (marianogonzalez) [inline extensions/ftp/src/test/java/org/mule/extension/ftp/FtpCopyTestCase.java]: nope... already explained before

File: modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileContentWrapper.java (hunk 14 above)
Comment (marianogonzalez) [inline modules/file-extension-common/src/main/java/org/mule/runtime/module/extension/file/api/FileContentWrapper.java]: done

File: extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java (hunk 8 above)
Comment (pablolagreca) [inline extensions/ftp/src/main/java/org/mule/extension/ftp/internal/ftp/command/FtpCommand.java]: Please add the TODO with the jira issue.

**Pull request merged**